src/python/pants/backend/shell/goals/test.py

```diff
@@ -99,10 +101,16 @@ async def test_shell_command(
         if result.process_result.exit_code == 0:
             break
 
+    extra_output: Snapshot | None = None
+    if results[-1].adjusted_digest != EMPTY_DIGEST:
```

Comment:
Presumably `test_subsystem.attempts_default` can never be less than 1 ?

Reply:
That is a critical assumption. I'll add some code to validate it.

Reply:
Added validation code.